### PR TITLE
:seedling: Convert pinned dependencies to probe

### PR DESCRIPTION
**Pull request by AdamKorcz**
#### What kind of change does this PR introduce?
feature

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
Currently, the Pinned Dependencies check does not include probes.

#### What is the new behavior (if this is a feature change)?**

This converts the Pinned Dependencies evaluation to probes.

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
NONE

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: checks/evaluation/pinned_dependencies.go [hunk 1]
```diff
@@ -316,21 +192,6 @@ func updatePinningResults(dependencyType checker.DependencyUseType,
...
-
-	return fmt.Sprintf("%s not pinned by hash", rr.Type)
-}
-
-func generateTextIncompleteResults(e checker.ElementError) string {
-	return fmt.Sprintf("Possibly incomplete results: %s", e.Err)
-}
-
 func generateOwnerToDisplay(gitHubOwned bool) string {
 	if gitHubOwned {
```
Comment (spencerschrock) [inline checks/evaluation/pinned_dependencies.go:196]: so this code now exists in two places (here and the probe). Not sure this needs to be dealt with right away, but long term might want to clean this up.

File: checks/evaluation/pinned_dependencies_test.go [hunk 2]
```diff
@@ -228,602 +229,180 @@ func Test_createScoreForGitHubActionsWorkflow(t *testing.T) {
...
+				NumberOfInfo: 1,
 			},
 		},
 		{
-			name: "GitHub Actions ecosystem with GitHub-owned and third-party unpinned",
-			dependencies: []checker.Dependency{
-				{
-					Location: &checker.File{
-						Snippet: "actions/checkout@v2",
+			name: "2 unpinned dependencies for 1 ecosystem shows 2 warn messages",
```
Comment (spencerschrock) [inline checks/evaluation/pinned_dependencies_test.go:348]: some sort of copy/paste error here.

1 ecosystem -> 2 ecosystem maybe?

File: checks/evaluation/pinned_dependencies_test.go [hunk 3]
```diff
@@ -852,35 +424,6 @@ func stringAsPointer(s string) *string {
 	return &s
 }
 
-func Test_generateOwnerToDisplay(t *testing.T) {
-	t.Parallel()
-	tests := []struct { //nolint:govet
-		name        string
-		gitHubOwned bool
-		want        string
-	}{
```
Comment (spencerschrock) [inline checks/evaluation/pinned_dependencies_test.go:861]: this test could probably stay if `generateOwnerToDisplay` is still in the code?

File: probes/pinsDependencies/impl_test.go [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+					},
+				},
+			},
+			outcomes: []finding.Outcome{
+				finding.OutcomePositive,
+				finding.OutcomePositive,
+			},
+		},
+		{
+			name: "GitHub Actions ecosystem with GitHub-owned and third-party pinned",
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl_test.go:378]: copy/paste on name?

pinned -> unpinned?

File: probes/pinsDependencies/impl_test.go [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+					},
+				},
+			},
+			outcomes: []finding.Outcome{
+				finding.OutcomeNegative,
+				finding.OutcomePositive,
+			},
+		},
+		{
+			name: "Skipped objects and dependencies",
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl_test.go:459]: what's skipped about this test?

File: probes/pinsDependencies/impl_test.go [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+						},
+					},
+				},
+			},
+			outcomes: []finding.Outcome{
+				finding.OutcomeNotApplicable,
+			},
+		},
+		{
+			name: "2 processing errors",
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl_test.go:552]: nit: describe the purpose, not describe the test.

```
name: "processing errors result in OutcomeError"
```

File: checks/evaluation/pinned_dependencies.go [hunk 7]
```diff
@@ -95,132 +75,28 @@ func probeRemToRuleRem(rem *probe.Remediation) *rule.Remediation {
...
-	findings, err := dependenciesToFindings(r)
-	if err != nil {
-		return checker.CreateRuntimeErrorResult(name, err)
-	}
 
 	for i := range findings {
 		f := findings[i]
 		switch f.Outcome {
+		case finding.OutcomeNotAvailable:
+			return checker.CreateInconclusiveResult(name, "no dependencies found")
```
Comment (spencerschrock) [inline checks/evaluation/pinned_dependencies.go:99]: optional: I think there are few mismatched outcomes -> behavior. But as a purely conversion PR this isn't necessarily in scope on this PR.

if you'd like to tackle it here, we should briefly discuss changes before switching them.

**Review by spencerschrock — APPROVED** [review]
Generally ok as a conversion PR. Left some comments

File: checks/evaluation/pinned_dependencies.go (hunk 7 above)
Comment (spencerschrock) [reply checks/evaluation/pinned_dependencies.go:99]: I think the `finding.OutcomeNotAvailable` and `finding.OutcomeNotApplicable` would be swapped?

File: probes/pinsDependencies/impl_test.go (hunk 5 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl_test.go:459]: I don't know. I on purpose did not change test names to ensure compatibility. This is the test from here: 

https://github.com/ossf/scorecard/blob/b21a350f500a1ec4552ab40a1a9af236cfc06d97/checks/evaluation/pinned_dependencies_test.go#L801-L828

File: probes/pinsDependencies/impl_test.go (hunk 6 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl_test.go]: https://github.com/ossf/scorecard/pull/3829/commits/a6947e6ca7b69a7153ab857cd34a3f1f2feb790f

File: probes/pinsDependencies/impl_test.go (hunk 4 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl_test.go]: https://github.com/ossf/scorecard/pull/3829/commits/2d2bb309032414225ed361e43c08245e1e01790b

File: checks/pinned_dependencies_test.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		files   []string
+		want    scut.TestReturn
+		wantErr bool
+	}{
+		{
+			name: "Dockerfile",
+			path: "./testdata/pinneddependencies/Dockerfile-script-ok",
+			files: []string{
+				"Dockerfile-script-ok",
+			},
```
[the remark this answers is not in the document]
Comment (AdamKorcz) [reply checks/pinned_dependencies_test.go]: Removed the copy here: https://github.com/ossf/scorecard/pull/3829/commits/17ea991fca89515520e7a19858acd7c6e6e3209c

File: checks/evaluation/pinned_dependencies_test.go (hunk 2 above)
Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies_test.go]: Thanks! https://github.com/ossf/scorecard/pull/3829/commits/dd07fede89c1a68fbf377d8a8c0b9db933700247

File: checks/evaluation/pinned_dependencies_test.go (hunk 3 above)
Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies_test.go:859]: Referencing https://github.com/ossf/scorecard/pull/3829#discussion_r1479946774

File: checks/evaluation/pinned_dependencies.go (hunk 7 above)
Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies.go:96]: Let's defer this to a later PR. I have created a tracker here: https://github.com/ossf/scorecard/issues/3855

File: checks/evaluation/pinned_dependencies.go (hunk 1 above)
Comment (spencerschrock) [reply checks/evaluation/pinned_dependencies.go:193]: i'd rather stick it in some `internal` folder if that's the goal.

Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies.go:193]: > honestly it's small enough, I think we can deal with the temporary duplication

Alright. Closing.

File: checks/evaluation/pinned_dependencies_test.go (hunk 3 above)
Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies_test.go:859]: Added here: https://github.com/ossf/scorecard/pull/3829/commits/153fb4ea71319542114f99871dfe9b436e7ead9d

**Comment by AdamKorcz** [comment]
@spencerschrock I have pulled in the latest main for this one. PTAL.

File: probes/pinsDependencies/impl.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+	r := raw.PinningDependenciesResults
+
+	for i := range r.ProcessingErrors {
+		e := r.ProcessingErrors[i]
+		f := finding.Finding{
+			Message:  generateTextIncompleteResults(e),
+			Location: &e.Location,
+			Outcome:  finding.OutcomeError,
+		}
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl.go:54]: here, and below, should use the various `finding.New` functions. This one doesn't set the `Probe`, so you get runtime errors like:
```
check runtime error: Pinned-Dependencies: internal error: invalid probe results
```

File: checks/evaluation/pinned_dependencies.go [hunk 10]
```diff
@@ -224,15 +113,15 @@ func PinningDependencies(name string, c *checker.CheckRequest,
 				lm.Remediation = probeRemToRuleRem(f.Remediation)
 			}
 			dl.Warn(lm)
-		case finding.OutcomeNotAvailable:
+		case finding.OutcomeError:
```
Comment (spencerschrock) [inline checks/evaluation/pinned_dependencies.go:116]: should probably have a test for this. either now or in followups.

File: probes/pinsDependencies/def.yml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+motivation: >
+  Pinned dependencies ensure that checking and deployment are all done with the same software, reducing deployment risks, simplifying debugging, and enabling reproducibility. They can help mitigate compromised dependencies from undermining the security of the project (in the case where you've evaluated the pinned dependency, you are confident it's not compromised, and a later version is released that is compromised).
+implementation: >
+  The probe works by looking for unpinned dependencies in Dockerfiles, shell scripts, and GitHub workflows which are used during the build and release process of a project. Special considerations for Go modules treat full semantic versions as pinned due to how the Go tool verifies downloaded content against the hashes when anyone first downloaded the module.
+outcome:
+  - For each of the last 5 releases, the probe returns OutcomePositive, if the release has a signature file in the release assets.
+  - For each of the last 5 releases, the probe returns OutcomeNegative, if the release does not have a signature file in the release assets.
+  - If the project has no releases, the probe returns OutcomeNotApplicable.
```
[the remark this answers is not in the document]
Comment (spencerschrock) [reply probes/pinsDependencies/def.yml:24]: not resolved? Or are we including this as part of #3855 ?

File: probes/pinsDependencies/impl.go [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+		if rr.Location == nil {
+			if rr.Msg == nil {
+				e := sce.WithMessage(sce.ErrScorecardInternal, "empty File field")
+				return findings, Probe, e
+			}
+			f := &finding.Finding{
+				Probe:   Probe,
+				Outcome: finding.OutcomeNotApplicable,
+				Message: *rr.Msg,
+			}
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl.go:69]: finding.New

File: probes/pinsDependencies/impl.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+				LineStart: &rr.Location.Offset,
+				LineEnd:   &rr.Location.EndOffset,
+				Snippet:   &rr.Location.Snippet,
+			}
+			f := &finding.Finding{
+				Probe:    Probe,
+				Outcome:  finding.OutcomeNotApplicable,
+				Message:  fmt.Sprintf("%s has empty Pinned field", rr.Type),
+				Location: loc,
+			}
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl.go:103]: finding.New

File: probes/pinsDependencies/impl.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+				LineStart: &rr.Location.Offset,
+				LineEnd:   &rr.Location.EndOffset,
+				Snippet:   &rr.Location.Snippet,
+			}
+			f := &finding.Finding{
+				Probe:    Probe,
+				Outcome:  finding.OutcomeNegative,
+				Message:  generateTextUnpinned(&rr),
+				Location: loc,
+			}
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl.go:120]: finding.New

File: probes/pinsDependencies/impl.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+				Path:      rr.Location.Path,
+				LineStart: &rr.Location.Offset,
+				LineEnd:   &rr.Location.EndOffset,
+				Snippet:   &rr.Location.Snippet,
+			}
+			f := &finding.Finding{
+				Probe:    Probe,
+				Outcome:  finding.OutcomePositive,
+				Location: loc,
+			}
```
Comment (spencerschrock) [inline probes/pinsDependencies/impl.go:140]: finding.New

Comment (AdamKorcz) [reply probes/pinsDependencies/impl.go]: https://github.com/ossf/scorecard/pull/3829/commits/1d58fbcca7933ce6f488fcb48e9178671ca79a9b

File: probes/pinsDependencies/impl.go (hunk 14 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl.go]: https://github.com/ossf/scorecard/pull/3829/commits/1d58fbcca7933ce6f488fcb48e9178671ca79a9b

File: probes/pinsDependencies/impl.go (hunk 13 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl.go]: https://github.com/ossf/scorecard/pull/3829/commits/a8bfdb5364e676d37ba5567df332b68d24c99c67

File: checks/evaluation/pinned_dependencies.go (hunk 10 above)
Comment (AdamKorcz) [reply checks/evaluation/pinned_dependencies.go:116]: > should probably have a test for this. either now or in followups.

If possible, let's do it in a follow-up.

File: probes/pinsDependencies/impl.go (hunk 9 above)
Comment (AdamKorcz) [reply probes/pinsDependencies/impl.go]: Should all be fixed.

**Comment by AdamKorcz** [comment]
@spencerschrock PTAL again.

File: probes/pinsDependencies/def.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (spencerschrock) [reply probes/pinsDependencies/def.yml:24]: Response was added on main thread. linking for convenience https://github.com/ossf/scorecard/pull/3829#issuecomment-1960324532

**Comment by spencerschrock** [comment]
> @spencerschrock PTAL again.

LGTM. Will merge on Monday (just in case theres anything we didn't catch, it gives us more time to catch it before the cron uses it)

**Pull request merged**